### PR TITLE
DM: add message agent

### PR DESCRIPTION
**Pull request by GMHDBJD**
- add message agent for OpreateWorker/OperateTask
- implement a simple synchronous request/response message mechanisms for temporary usage.

### CODE REVIEW

**Comment by GMHDBJD** [comment]
cc @lance6716

File: pkg/dm/message_impl.go [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+func NewMessageImpl() *MessageImpl {
+	return &MessageImpl{
+		idAllocator: &MessageIDAllocator{},
+	}
+}
+
+// SendMessage sends a message asynchronously.
+func (m *MessageImpl) SendMessage(ctx context.Context, topic p2p.Topic, message Message, sender Sender) error {
```
Comment (lance6716) [inline pkg/dm/message_impl.go:46]: maybe this method is not needed, we have too many `SendMessage` now

File: pkg/dm/message_impl.go [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+}
+
+type Sender interface {
+	SendMessage(ctx context.Context, topic p2p.Topic, message interface{}, nonblocking bool) error
+}
+
+// MessageImpl implement a simple synchronous request/response message pattern since the lib currently only support asynchronous message.
+// Caller should persist the request message if needed.
+// Caller should add retry mechanism if needed.
+type MessageImpl struct {
```
Comment (lance6716) [inline pkg/dm/message_impl.go:32]: maybe MessageRecorder, MessageRouter or something. Message makes me think of a message entity that has NewMessage and some (de)serialization method.

File: pkg/dm/topic.go [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+	"github.com/hanfei1991/microcosm/pkg/p2p"
+)
+
+// Message use for asynchronous message.
+type Message interface{}
+
+// MessageWithID use for synchronous request/response message.
+type MessageWithID struct {
+	ID      uint64
+	Message interface{}
```
Comment (lance6716) [inline pkg/dm/topic.go:17]: the type should also be Message, as defined above

File: pkg/dm/message_impl.go [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+// OnResponse receives a response message.
+func (m *MessageImpl) OnResponse(msg MessageWithID) error {
+	respCh, ok := m.pendings.Load(msg.ID)
+	if !ok {
+		return errors.Errorf("request %d not found", msg.ID)
+	}
+
+	select {
+	case respCh.(chan Response) <- msg.Message:
+		m.pendings.Delete(msg.ID)
```
Comment (lance6716) [inline pkg/dm/message_impl.go:78]: I think we don't need this line (line 55)

File: jobmaster/dm/message_agent.go [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+	if _, ok := agent.senders.Load(taskID); ok {
+		return "", errors.Errorf("worker for task %s already exist", taskID)
+	}
+	return agent.masterImpl.CreateWorker(workerType, taskCfg, 1)
+}
+
+func (agent *MessageAgent) DestroyWorker(ctx context.Context, taskID lib.WorkerID, workerID lib.WorkerID) error {
+	v, ok := agent.senders.Load(taskID)
+	if !ok {
+		return errors.Errorf("worker for task %s not exist", taskID)
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply jobmaster/dm/message_agent.go:77]: Let outer logic to handle the error.

File: pkg/dm/topic.go [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+type Request Message
+
+type Response Message
+
+func OperateTaskMessageTopic(masterID lib.MasterID, taskID string) p2p.Topic {
+	return fmt.Sprintf("operate-task-message-%s-%s", masterID, taskID)
+}
+
+type OperateTaskMessage struct {
+	TaskID string
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply pkg/dm/topic.go:29]: I think it is better to repeat it in payload, do not need to known the specific format of the topic.

File: jobmaster/dm/message_agent.go [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+		TaskID: taskID,
+		Stage:  stage,
+	}
+
+	ctx, cancel := context.WithTimeout(ctx, DefaultMessageTimeOut)
+	defer cancel()
+	return agent.messageImpl.SendMessage(ctx, topic, message, v.(Sender))
+}
+
+func (agent *MessageAgent) OnWorkerMessage(message interface{}) error {
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply jobmaster/dm/message_agent.go:115]: This actually is master-worker relation, the MessageAgent hold by master(DM-master), send message to worker(DM-worker)

File: jobmaster/dm/worker_manager.go [hunk 8]
```diff
@@ -264,3 +264,12 @@ func (wm *WorkerManager) destroyWorker(ctx context.Context, taskID string, worke
 	wm.workers.Delete(taskID)
 	return nil
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply jobmaster/dm/worker_manager.go:265]: I think it's better to separate, each component maintains its own state.

File: pkg/dm/message_impl.go (hunk 1 above)
Comment (GMHDBJD) [reply pkg/dm/message_impl.go:46]: We need sends a message asynchronously, caller(message_agent) needs to decide for itself whether a message is asynchronous or not.

File: pkg/dm/message_impl.go (hunk 2 above)
Comment (GMHDBJD) [reply pkg/dm/message_impl.go:32]: I change it to MessagePair...

File: jobmaster/dm/message_agent.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+	// for create worker
+	CreateWorker(workerType lib.WorkerType, config lib.WorkerConfig, cost model.RescUnit) (lib.WorkerID, error)
+	// for operate-task
+	CurrentEpoch() lib.Epoch
+	// for topic
+	JobMasterID() lib.MasterID
+}
+
+type Sender interface {
+	ID() lib.WorkerID
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply jobmaster/dm/message_agent.go:31]: message agent hold by master, so the sender it is a worker agent.

[the remark this answers is not in the document]
Comment (lance6716) [reply jobmaster/dm/message_agent.go:31]: My understand for

> message agent hold by master, so the sender it is a worker agent.

is that the `Sender` interface is like a client/agent. Master holds many `Sender`, each for a worker, so the ID represent the workerID.

My first thought when see it is that it's the sender of sender/receiver pair, so I'm expected its ID is the masterID. Now I prefer we name this interface like SendHandle/SendAgent....

And, can worker start a sending?

File: pkg/dm/message_impl.go (hunk 1 above)
Comment (lance6716) [reply pkg/dm/message_impl.go]: I guess your meaning is when dealing with messages, we should always construct a MessagePair, and use it to synchronously or asynchronously send messages by `SendMessage` or `SendRequest`. 

Both `SendMessage` or `SendRequest` need the actual message and a sender as parameters, for `SendMessage` caller can directly use sender to send the message, no need to construct a MessagePair. This is also stated by the definition of MessagePair: "MessagePair implement a simple synchronous request/response message pattern". And the caller (I guess) will not dynamically choose synchronously or asynchronously sending, most method should only need one of them by the semantic.

For these reason I think wrap a `SendMessage` to MessagePair is not needed.

File: jobmaster/dm/message_agent.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply jobmaster/dm/message_agent.go:31]: Sender is a interface wrap with WorkerHandle, if worker want to sending, it should write its own code(message_agent) with it's sender.

File: jobmaster/dm/message_agent.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+	SendMessage(ctx context.Context, topic string, message interface{}, nonblocking bool) error
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply jobmaster/dm/message_agent.go:32]: add comment in topic.go

File: pkg/dm/message_impl.go (hunk 1 above)
Comment (GMHDBJD) [reply pkg/dm/message_impl.go]: removed

File: jobmaster/dm/message_agent.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply jobmaster/dm/message_agent.go:31]: just like master grpc proto and worker grpc proto

**Review by lance6716 — COMMENTED** [review]
rest lgtm

**Review by lance6716 — APPROVED** [review]

**Comment by GMHDBJD** [comment]
cc @liuzix @amyangfei @nongfushanquan

File: pkg/dm/message_pair.go [hunk 11]
```diff
@@ -0,0 +1,77 @@
...
+	if !ok {
+		return errors.Errorf("request %d not found", msg.ID)
+	}
+
+	select {
+	case respCh.(chan Response) <- msg.Message:
+		return nil
+	default:
+	}
+	return errors.Errorf("duplicated response of request %d, and the last response is not consumed", msg.ID)
```
Comment (amyangfei) [inline pkg/dm/message_pair.go:76]: Does this mean the message responser guarantees that the message is acked at most once.

File: jobmaster/dm/message_agent.go [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+}
+
+// MessageAgent hold by Jobmaster, it manage all interactions with workers
+type MessageAgent struct {
+	master      Master
+	clocker     clock.Clock
+	messagePair *dmpkg.MessagePair
+	// for stop a task
+	id lib.WorkerID
```
[the remark this answers is not in the document]
Comment (GMHDBJD) [reply jobmaster/dm/message_agent.go:41]: fix, it's the id of dm_jobmaster.

File: pkg/dm/message_pair.go (hunk 11 above)
Comment (GMHDBJD) [reply pkg/dm/message_pair.go:76]: yes

**Pull request merged**